marshmallow/fields.py

```diff
@@ -803,17 +805,21 @@ def _deserialize(self, value):
         err = ValidationError(getattr(self, 'error', None) or msg)
         if not value:  # Falsy values, e.g. '', None, [] are not valid
             raise err
-        self.dateformat = self.dateformat or self.DEFAULT_FORMAT
         func = self.DATEFORMAT_DESERIALIZATION_FUNCS.get(self.dateformat)
         if func:
             try:
                 return func(value)
             except (TypeError, AttributeError, ValueError):
                 raise err
+        elif type(value) is str and self.dateformat:
```

Comment:
I'm not sure that the typechecking is necessary. We can assume that `self.dateformat` is either a string or `None`.